### ISSUE TITLE
Rust Binding broken

### ISSUE THREAD

**Issue by kodebach** [ElektraInitiative/libelektra#4781]
The Rust binding is broken. You can see this in (at least) #4778, #4717 and #4715. A rerun does not help.

https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/PR-4778/5/pipeline/
https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/PR-4715/9/pipeline/
https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/PR-4717/7/pipeline/

The issue seems to be that the we use the `rustc` on Debian:

https://github.com/ElektraInitiative/libelektra/blob/90f70811c51a76e758f671f373cace2b98b412ed/scripts/docker/debian/bullseye/Dockerfile#L73

**THIS IS A VERY BAD IDEA**

The Rust version packaged by Debian (1.48.0 2020-11-19) is basically ancient compared to the current on (1.66.0 2022-12-17). AFAICT the issue is that one of the libraries that we transitively depend on no longer works with this old version (see also https://github.com/gimli-rs/object/issues/394).

There might be another way to fix this, but the easiest way would be to change your CI config & and Dockerfiles to install Rust the recommended way, via `rustup` (https://www.rust-lang.org/tools/install). Then we could use the current version, which builds the bindings without issue.

In any case, we need a quick decision, because AFAICT this breaks all Jenkins builds

cc @markus2330 @lukashartl @flo91 @mpranj

**Comment by markus2330** [comment]
Thank you for reporting this urgent problem!

One problem is that the compilation of the dependency object-0.30.0 fails. Trying to specify object 0.29.0 didn't help for me, as backtrace-0.3.67, required by bindgen- 0.50.1 depends on object-0.30.0. So I fixed the version of bindgen. Seems to be a semver fail, or at least in Rust's interpretation of it. https://semver.org/ itself says "Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable." so it is actually wrong to allow 0.50.1 as substitute for 0.50.0.

Anyway, I fully agree that we *also* should test rustup. Using rustup has one more advantage: then we would have rustfmt available for code formatting.

But when maintainers compile Elektra for their distribution they cannot use rustup, and I think we should also care for them, especially as long our Rust code compiles with older compilers. (If we want later Rust features it is a different story.)

> Then we could use the current version, which builds the bindings without issue.

Did you try with Debian Bullseye? For me using 1.66.0 did not solve the issue (could be a problem with parallel installation of cargo, though) but #4782 worked for me with both 1.46.0 and 1.66.0 locally. On the CI, however, #4782 fails nevertheless?

Help appreciated.

@PhilippGackstatter can you maybe take a look?

**Comment by PhilippGackstatter** [comment]
Sorry didn't see this earlier. Was about to post an answer, but glad to see you figured out a fix.

What I was basically going to say is that upgrading to the latest rust version using rustup is probably going to be the most comprehensive fix and will cause much fewer problems in the future, too.

**Comment by markus2330** [comment]
Thanks for the answer! Yes, we are using rustup now for our main image.

**Comment by kodebach** [comment]
> [...] so it is actually wrong to allow 0.50.1 as substitute for 0.50.0.

True, but AFAICT it's quite common to interpret 0.x.y to 0.x.z as a non-breaking change and only 0.x.y to 0.a.b as breaking.

> Did you try with Debian Bullseye? [...]

I tried only on Manjaro. But the default 1.66.0 version worked and when I switched to 1.48.0 via rustup I got the same error as the CI.

I also agree that we should test the version used in Debian packages, but IMO we should still install with rustup and just specify the version explicitly with e.g. `cargo +1.48.0 build`. (Not sure how to do that via our CMake setup right now)
